mfs: Call int2f/120c for FCBs with DOS Stack #928

andrewbird opened the pull request:
With this change Castle Wolfenstein doesn't complain about the control
file not opening and now reaches the splash screen.

I don't play games myself so I've no idea what's supposed to happen next. I pressed `<return>` but ended up with a blank screen. I see something similar reported in #920, but it'd be good if someone else could look this over, please?

stsp commented:
But do you know why exactly is this needed?
What does this change?

stsp commented:
Yeah, blank screen is quite fine.
But it would be better to know the magic
behind this change.

andrewbird commented:
I wrote a quick extension to dosdebug to display an SFT entry at addr and could see that the original calling method was depositing rubbish in some of the SFT fields. I didn't check but it may have been setting an error field somewhere in there too. Switching to my current method gave what looked like proper values. Of course that doesn't explain why the old method doesn't work, but I didn't see why int2f/120c was called differently to other int2f/12xx functions we have used in mfs.c, so I switched over without too much investigation.

I should probably add a couple of tests for FCBopen/read & FCBopen/write too.

stsp commented:
Usually this way of calling is used if you
immediately need, right in the same C code,
the result of that call. If you don't need the
result immediately and only want it to be
returned to the user, the previous way should
do too.
If it doesn't work that way, then the problem
is possibly elsewhere. I'll take a look myself
at the place you pointed, to see what's the
difference. But right now my PC is burned,
so you have enough time to take a look yourself. :)

andrewbird commented:
Just writing fcb open/read/write tests at the moment.

andrewbird commented:
Here's the test for FCB open/read file https://github.com/andrewbird/dosemu2/commit/dc3a41576e5065179f489d1b847d7d280fad29b0
As you noted FDPP passes with current devel even on MFS, whereas FreeDOS 1.20 fails.

BTW I can see why MS invented the file handle access, FCB programming is awful.

Write test coming later.

stsp commented:
Will these tests show the need for the
change in question?

andrewbird commented:
Not without FreeDOS Test case added, but adding that adds other failures though- see https://travis-ci.com/andrewbird/dosemu2/builds/135242498. FDPP passes without error on devel.

stsp commented:
You probably need to differentiate the freedos
failure and dosemu failure. You fixed most of
those in fdpp, so you know for sure which are
_definitely_ the freedos failures, and which are
not. For those that are, simply don't print "FAIL" -
print something else for it to actually "pass".

andrewbird commented:
I've now added the ability to skip certain tests based on the TestCase instance. I also added a FCB write file test. Both that and the FCB read file test are failing on FreeDOS 1.20 with MFS, but okay on FAT. See https://travis-ci.com/andrewbird/dosemu2/builds/135493999

stsp commented:
This is one way of doing it, but its probably
not the best one because if you just skip the
test, you won't notice if it ever got fixed.
So I would suggest some other result, which
will mean "skipped because known-bad test failed".
You probably would like also another one status:
"wow, known-bad test _passed_" :)

andrewbird commented:
Yes, that's what hoped to do, but python's unittest framework doesn't provide a `TestCase.expectedFailure()` method, only a decorator. So to achieve it I'd have to do this for each test (on FreeDOS 1.20 there are currently 11 tests to do this for). Perhaps there's a better way, anyone?

``` Python
class FRDDOS120TestCase(BaseTestCase):

    @unittest.expectedFailure
    def test_mfs_fcb_read(self):
        super(FRDDOS120TestCase, self).test_mfs_fcb_read()
```

stsp commented:
How many freedos-1.20 tests in total?
How is it possible to not do that per
each needed test, if you cartainly don't
want to do that for all freedos-1.20 tests?
Some kind of grouping?
How does this expactedFailure would
react on the test eventually passing?
Just a newbie questions, never used
that stuff.

andrewbird commented:
72 I think, of which 11 need the decorator applied. I added a test commit, but I'm not happy with the way the decorator loses the test name. Also I think I'll need to switch to python 3, as the test runner on Travis doesn't know how to deal epectedfailures. The proper behaviour is that tests marked expectedfailure return fail if they pass, and return pass if they fail. More work required...

stsp commented:
Fail if they pass, doesnt look good,
at least not any better than skip.
If there is no some "mayfail", lets just
use skip...

andrewbird commented:
I think the return fail if they passed (and were marked expectedFailure) is to alert the tester that something changed, anyway it's far easier/neater for me to just drop the last commit. When next release of FreeDOS lands we can use that and most, if not all, will go away

stsp commented:
Ok, fine with skip then.
Why do you think the problems will
be fixed on the next release?
The amount of freedos bugs we
fixed, will likely take eternity on their
side.

andrewbird commented:
> Why do you think the problems will
be fixed on the next release?

Only that the failures are mostly FCB related and at least some are fixed by https://github.com/FDOS/kernel/commit/d5b5505013c9a53d93baa1ae3cde541cbd25c15e .

Locally I've tested FreeDOS from git (built using ia16-gcc) with these same tests but have varying results.

stsp commented:
Thanks for debugging this.
The problem appears trivial (knowing where to look
is most important of course).
